DirectProgramming/DPC++FPGA/Tutorials/DesignPatterns/shannonization/src/shannonization.cpp

```diff
@@ -0,0 +1,335 @@
+  int iterations = 5;
+#endif
+
+  std::cout << "Running " << iterations 
+            << ((iterations == 1) ? " iteration" : " iterations")
+            << " of kernel " << Version
+            << " with |A|=" << a.size() 
+            << " and |B|=" << b.size() << "\n";
+  
+  bool success = true;
+  std::vector<double> kernel_latency(iterations);
+
+  // perform multiple iterations of the kernel to get a more accurate
+  // throughput measurement
+  for (size_t i = 0; i < iterations && success; i++) {
+    // run kernel
+    int n = 0;
+    event e = SubmitKernels<Version,II>(q, a, b, n);
+
+    // check output
+    if (golden_n != n) {
+      success = false;
+      std::cerr << "ERROR: Kernel version " << Version << " output is incorrect"
+                << " (Expected=" << golden_n << ", Result=" << n << ")\n";
+    }
+
+    // get profiling info
+    auto start = e.get_profiling_info<info::event_profiling::command_start>();
+    auto end = e.get_profiling_info<info::event_profiling::command_end>();
+    kernel_latency[i] = (end - start) / 1000000000.0f;
+  }
+
+  // If all the iterations were successful, print the throughput results.
+  // The FPGA emulator does not accurately represent the hardware performance
+  // so we don't print performance results when running with the emulator
+  if(success) {
+#ifndef FPGA_EMULATOR
+    // Compute the average throughput across all iterations.
+    // We use the first iteration as a 'warmup' for the FPGA,
+    // so we ignore its results.
+    double avg_kernel_latency =
+        std::accumulate(kernel_latency.begin() + 1, kernel_latency.end(), 0.0) /
+        (double)(iterations - 1);
+
+    double input_size_megabytes = 
+        ((a.size() + b.size()) * sizeof(unsigned int)) / (1024.0 * 1024.0);
+
+    const double avg_throughput = input_size_megabytes / avg_kernel_latency;
+
+    std::cout << "Kernel " << Version 
+              << " average throughput: " << avg_throughput << " MB/s\n";
+#endif
+  }
+
+  return success;
+}
+
+
+int main(int argc, char** argv) {
+  // parse the command line arguments
+#if defined(FPGA_EMULATOR)
+  unsigned int a_size = 128;
+  unsigned int b_size = 256;
+#else
+  unsigned int a_size = 131072;
+  unsigned int b_size = 262144;
+#endif
+  bool need_help = false;
+
+  for (int i = 1; i < argc; i++) {
+    std::string arg(argv[i]);
+
+    if (arg == "--help" || arg == "-h") {
+      need_help = true;
+    } else {
+      std::string str_after_equals = arg.substr(arg.find("=") + 1);
+
+      if (StrStartsWith(arg, "--A=")) {
+        a_size = std::stoi(str_after_equals);
+      } else if (StrStartsWith(arg, "--B=")) {
+        b_size = std::stoi(str_after_equals);
+      } else {
+        std::cout << "WARNING: ignoring unknown argument '" << arg << "'\n";
+      }
+    }
+  }
+
+  // ensure the arrays have more than 3 elements
+  if(a_size <= 3) {
+    std::cout << "WARNING: array A must have more than 3 "
+                  "elements, increasing its size\n";
```

Comment:
For `std::cout` use `std::endl` instead of `\n` everywhere. Otherwise you might never see the output if the program crashes...

Reply:
I don't think this is true. If the program catches, the exception handler will call std::terminate() which should flush the buffer. Also, the use of \n rather than std::endl is mandated by the coding standard for code samples. I believe that if an explicit flush is desired (of which I'm not convinced here) that an explicit std::flush is preferred.

Reply:
Our internal guidelines say to do the exact opposite. @akertesz ??

Reply:
> I don't think this is true. If the program catches, the exception handler will call std::terminate() which should flush the buffer.

Can you show me where this is stated in the C++ standard?

> Also, the use of \n rather than std::endl is mandated by the coding standard for code samples. 

Strange... @erichkeane I am curious about who writes them...

Reply:
In cout, \n causes a flush as far as I know, which is why the coding  standard recommends doing  '\n'. As far as std::terminate flushing the buffer, I'm not aware of that being a standards requirement, but I can look closer tomorrow.

Reply:
No emergency. :-) I think `\n` causes a flush in C, not in C++... Perhaps this is the source of the confusion in the coding guideline?

Reply:
The '\n' flushing for cout is an implementation behavior that all major ones share.  

See here for ifo about the terminate flushing: https://en.cppreference.com/w/cpp/io/manip/flush

"In most other usual interactive I/O scenarios, std::endl is redundant when used with std::cout because any input from std::cin, output to std::cerr, or program termination forces a call to std::cout.flush(). "

Reply:
Yes but none of these applies. In the case of a crash or `std::terminate` (aka `std::abort()` on Unix by default!), there is no flush.
Since nobody believes me or https://isocpp.org/wiki/faq/input-output#endl-vs-slash-n, here is the proof:
https://godbolt.org/z/68GWsE

Reply:
The problem with coding guide-lines for **tutorials** is that there should not inoculate more bad coding practices in customer code... Life is already hard enough like that... :-/

Reply:
Based on our internal guidelines, which state:
> For line endings, use '\n' over std::endl (unless you actually need the explicit flush as well as the newline).

I believe the compromise here would be to continue using '\n' BUT place a `std::flush()` before the call to `std::terminate`.

What do you think @akertesz and @keryell??

Reply:
Yes, that was my proposal. 

Any exception in the tutorial code will be handled by either the asynchronous handler, or caught by the synchronous try-catch block. I just checked the asynchronous handler in dpc_common.hpp, and saw that it uses std::endl. If we add an explicit flush in the synchronous catch block before the std::terminate, we're covered.

Personally I prefer std::flush to std::endl as it is more explicit, but probably the correct thing is to follow the style in dpc_common.hpp.

Reply:
I believe that I am right now in a TV reality show about code obfuscation.
Are you SERIOUS??? `\n` + `std::flush`?
Why not copy-pasting the standard C++ library to be sure we are explicit... ;-)
And add a few more `std::flush` just in case... :-)

Reply:
I am no code standard guru, but having `std::endl` all over the place seems a little verbose, therefore, I am a fan of `\n`. That being said, I think in the case of the catch block, perhaps we should be using `std::cerr` in the catch block? That way the messages are flushed?

Reply:
Yes, `std::cerr` is the way to go here anyway.